open_lm/train.py

```diff
@@ -132,14 +134,32 @@ def train_one_epoch(model, data, loss, epoch, step, optimizer, scaler, scheduler
 
     end = time.time()
 
+    try:
+        starting_step = int(optimizer.state_dict()["state"][0]["step"].item())
+    except KeyError:
+        # Throws keyerror if it is the first step
+        starting_step = 0
+    done_training = False
+
+    other_proc_ran_out_of_data = False
     for i, batch in enumerate(dataloader):
         if not args.skip_scheduler:
             scheduler(step)
 
-        if step > total_steps:
+        if step >= total_steps:
             logging.warning(f"step: {step} exceeds total_steps: {total_steps}. ending training.")
             break
 
+        # If some process no longer has data, stop training - for this epoch
+        procs_with_data = torch.tensor(1, dtype=torch.long, device=device)
+        dist.all_reduce(procs_with_data, op=ReduceOp.SUM)
+        if procs_with_data < args.world_size:
+            other_proc_ran_out_of_data = True
+            break
+
+        if not args.skip_scheduler:
```

Comment:
I think this is a copy paste error - we call scheduler(step) above on L147.

Reply:
My bad - removed this.